service/src/main/java/bio/terra/workspace/service/resource/controlled/flight/clone/workspace/CloneEachResourceStep.java

```diff
@@ -0,0 +1,323 @@
+import bio.terra.workspace.service.workspace.model.WsmCloneResourceResult;
+import bio.terra.workspace.service.workspace.model.WsmResourceCloneDetails;
+import com.fasterxml.jackson.core.type.TypeReference;
+import java.time.Duration;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.UUID;
+import java.util.stream.Collectors;
+import javax.annotation.Nullable;
+import org.springframework.http.HttpStatus;
+
+public class CloneEachResourceStep implements Step {
+
+  public static final Duration FLIGHT_POLL_INTERVAL = Duration.ofSeconds(10);
+  public static final int FLIGHT_POLL_CYCLES = 100;
+  private final ReferencedResourceService referencedResourceService;
+  private final ControlledResourceService controlledResourceService;
+
+  public CloneEachResourceStep(
+      ReferencedResourceService referencedResourceService,
+      ControlledResourceService controlledResourceService) {
+    this.referencedResourceService = referencedResourceService;
+    this.controlledResourceService = controlledResourceService;
+  }
+
+  @Override
+  public StepResult doStep(FlightContext context) throws InterruptedException, RetryException {
+    final var userRequest =
+        context
+            .getInputParameters()
+            .get(JobMapKeys.AUTH_USER_INFO.getKeyName(), AuthenticatedUserRequest.class);
+    final var destinationWorkspaceId =
+        context.getWorkingMap().get(ControlledResourceKeys.DESTINATION_WORKSPACE_ID, UUID.class);
+    final var location =
+        context.getInputParameters().get(ControlledResourceKeys.LOCATION, String.class);
+    // get the list
+    final List<WsmResource> resourcesToClone =
+        context
+            .getWorkingMap()
+            .get(ControlledResourceKeys.RESOURCES_TO_CLONE, new TypeReference<>() {});
+    // Get or create the map from entry to result. This will be incrementally upgraded.
+    final Map<UUID, WsmResourceCloneDetails> resourceIdToResult =
+        Optional.ofNullable(
+                context
+                    .getWorkingMap()
+                    .get(
+                        ControlledResourceKeys.RESOURCE_ID_TO_CLONE_RESULT,
+                        new TypeReference<Map<UUID, WsmResourceCloneDetails>>() {}))
+            .orElseGet(HashMap::new);
+
+    // Clone first resource on the list that isn't in the map.
+    final Optional<WsmResource> resourceMaybe =
+        resourcesToClone.stream()
+            .filter(r -> !resourceIdToResult.containsKey(r.getResourceId()))
+            .findFirst();
+    if (resourceMaybe.isPresent()) {
+      final WsmResource resource = resourceMaybe.get();
+      try {
+        final WsmResourceCloneDetails cloneDetails =
+            clone(resource, destinationWorkspaceId, userRequest, location, context.getStairway());
+        resourceIdToResult.put(resource.getResourceId(), cloneDetails);
```

Comment:
I'm OK with it. There's not much of this class that I'll miss.

There are other places where I use the service methods instead of proper sub-flights, captured in PF-918.